### PR TITLE
fix: enter animations with mountOnEnter or unmountOnExit 

### LINKED ISSUE #745: Transition enter animation doesn't work correctly with unmountOnExit
> What is the current behavior?

When using the `Transition` component with `unmountOnExit`, the exit animation works correctly, but the enter animation doesn't work. However, when I remove `unmountOnExit`, the enter animation works.

> What is the expected behavior?

When using the `Transition` component with `unmountOnExit`, I expect the enter animation to work correctly.

> Could you provide a [CodeSandbox](https://codesandbox.io/) demo reproducing the bug?

https://codesandbox.io/s/admiring-fog-lhroc?file=/FadeScaleAlert.js

### PR DESCRIPTION
**Pull request by koba04**
fixes #745 
This fixes enter animations with `mountOnEnter` or `unmounOnExit`. However, this might introduce other issues like unexpected animations, so I have to test this before this gets merged.

### CODE REVIEW

**Comment by koba04** [comment]
I've added more stories for CSSTransition and it seems to work fine.
I've found some issues around TransitionGroup and couldn't figure out whether Transition should mount an element with unmountOnExit and without mountOnEnter or not. Currently, it's not mounted.
But I think this is another issue, so I won't fix it in this issue.

**Comment by koba04** [comment]
This introduces an additional render because `performEnter` is no longer batched.

**Comment by anatolzak** [comment]
Hi @koba04.

Thanks again for working on this pull request to fix the bug that I reported.

Is there perhaps a timeline to when this pull request will be merged into the main branch?

**Comment by koba04** [comment]
@anatolzak 
I don't have any timeline to merge this yet because it has a risk to cause other problems caused by an additional render to make this animation happen.
But I'll try to work to be able to merge this.

**Comment by ThomasEsseul** [comment]
Hi, any update on that ? I'm really interested in that fix

**Comment by koba04** [comment]
I'll check it again.

**Comment by koba04** [comment]
I've reworked this to get merged.

File: stories/CSSTransition.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
+import React, { useState } from 'react';
+import { storiesOf } from '@storybook/react';
+
+import StoryFixture from './StoryFixture';
+import Fade from './transitions/CSSFade';
+
+function ToggleFixture({ defaultIn, description, children }) {
+  const [show, setShow] = useState(defaultIn || false);
+  console.log(show);
```
Comment (koba04) [inline stories/CSSTransition.js:9]:
```suggestion
```

**Comment by koba04** [comment]
@eps1lon Could you review this PR?

**Comment by danantal** [comment]
Hi guys

I am also pretty interested in this fix - so would be amazing if it can be moved further :)

**Comment by koba04** [comment]
I think I had enough time to get feedback, so I'll merge this.

**Pull request merged**